### ISSUE TITLE
Allow unprivileged OpenLDAP LXC container

### ISSUE THREAD

**Issue by IzaacJ** [turnkeylinux/tracker#1535]
repurposed issue to track support for unprivileged OpenLDAP LXC container on Proxmox

----

Hello.
Im using Proxmox and an lxc container with turnkey-openldap 16.0-1 template, and stunnel fails to start, which in turn results in webmin failing to start.

This is after a fresh install, just after completing the initial setup and exiting from the config utility.

```
root@ldap ~# cat /etc/turnkey_version
turnkey-openldap-16.0-buster-amd64
root@ldap ~# netstat -tulpn
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name    
tcp        0      0 127.0.0.1:25            0.0.0.0:*               LISTEN      506/master          
tcp        0      0 0.0.0.0:443             0.0.0.0:*               LISTEN      1458/lighttpd       
tcp        0      0 0.0.0.0:636             0.0.0.0:*               LISTEN      2518/slapd          
tcp        0      0 0.0.0.0:389             0.0.0.0:*               LISTEN      2518/slapd          
tcp        0      0 0.0.0.0:80              0.0.0.0:*               LISTEN      1458/lighttpd       
tcp        0      0 0.0.0.0:22              0.0.0.0:*               LISTEN      1559/sshd           
tcp6       0      0 :::636                  :::*                    LISTEN      2518/slapd          
tcp6       0      0 :::389                  :::*                    LISTEN      2518/slapd          
tcp6       0      0 :::80                   :::*                    LISTEN      1458/lighttpd       
tcp6       0      0 :::22                   :::*                    LISTEN      1559/sshd           
root@ldap ~# service webmin status
* webmin.service - Webmin Web based Admin UI
   Loaded: loaded (/lib/systemd/system/webmin.service; enabled; vendor preset: enabled)
  Drop-In: /etc/systemd/system/webmin.service.d
           `-override.conf
   Active: inactive (dead)

Nov 10 21:55:41 ldap systemd[1]: Dependency failed for Webmin Web based Admin UI.
Nov 10 21:55:41 ldap systemd[1]: webmin.service: Job webmin.service/start failed with result 'dependency'.
root@ldap ~# service webmin start
A dependency job for webmin.service failed. See 'journalctl -xe' for details.
root@ldap ~# journalctl -xe
-- Automatic restarting of the unit stunnel4@webmin.service has been scheduled, as the result for
-- the configured Restart= setting for the unit.
Nov 10 21:57:43 ldap systemd[1]: Stopped Universal SSL tunnel for network daemons (webmin).
-- Subject: A stop job for unit stunnel4@webmin.service has finished
-- Defined-By: systemd
-- Support: https://www.debian.org/support
-- 
-- A stop job for unit stunnel4@webmin.service has finished.
-- 
-- The job identifier is 1617 and the job result is done.
Nov 10 21:57:43 ldap systemd[1]: stunnel4@webmin.service: Start request repeated too quickly.
Nov 10 21:57:43 ldap systemd[1]: stunnel4@webmin.service: Failed with result 'exit-code'.
-- Subject: Unit failed
-- Defined-By: systemd
-- Support: https://www.debian.org/support
-- 
-- The unit stunnel4@webmin.service has entered the 'failed' state with result 'exit-code'.
Nov 10 21:57:43 ldap systemd[1]: Failed to start Universal SSL tunnel for network daemons (webmin).
-- Subject: A start job for unit stunnel4@webmin.service has failed
-- Defined-By: systemd
-- Support: https://www.debian.org/support
-- 
-- A start job for unit stunnel4@webmin.service has finished with a failure.
-- 
-- The job identifier is 1617 and the job result is failed.
```

I tried the updated `stunnel4@.service` found in commit turnkeylinux/common@5cd698673671dea499e6af9f72dac69e5c93c4f5 but that does not work either, and renders this:

```
root@ldap ~# service webmin start
A dependency job for webmin.service failed. See 'journalctl -xe' for details.
root@ldap ~# journalctl -xe
-- Automatic restarting of the unit stunnel4@webmin.service has been scheduled, as the result for
-- the configured Restart= setting for the unit.
Nov 10 22:01:54 ldap systemd[1]: Stopped Universal SSL tunnel for network daemons (webmin).
-- Subject: A stop job for unit stunnel4@webmin.service has finished
-- Defined-By: systemd
-- Support: https://www.debian.org/support
-- 
-- A stop job for unit stunnel4@webmin.service has finished.
-- 
-- The job identifier is 1900 and the job result is done.
Nov 10 22:01:54 ldap systemd[1]: stunnel4@webmin.service: Start request repeated too quickly.
Nov 10 22:01:54 ldap systemd[1]: stunnel4@webmin.service: Failed with result 'exit-code'.
-- Subject: Unit failed
-- Defined-By: systemd
-- Support: https://www.debian.org/support
-- 
-- The unit stunnel4@webmin.service has entered the 'failed' state with result 'exit-code'.
Nov 10 22:01:54 ldap systemd[1]: Failed to start Universal SSL tunnel for network daemons (webmin).
-- Subject: A start job for unit stunnel4@webmin.service has failed
-- Defined-By: systemd
-- Support: https://www.debian.org/support
-- 
-- A start job for unit stunnel4@webmin.service has finished with a failure.
-- 
-- The job identifier is 1900 and the job result is failed.
```

Also tried the modified `webmin.service` found mentioned in issue #1480 together with the changed `stunnel4@.service` without any luck. It does however change the error a lot:

```
root@ldap ~# systemctl daemon-reload
root@ldap ~# systemctl restart webmin.service
A dependency job for webmin.service failed. See 'journalctl -xe' for details.
root@ldap ~# journalctl -xe
-- The process /usr/bin/stunnel4 could not be executed and failed.
-- 
-- The error number returned by this process is ERRNO.
Nov 10 22:11:10 ldap systemd[1]: stunnel4@webmin.service: Control process exited, code=exited, status=226/NAMESPACE
-- Subject: Unit process exited
-- Defined-By: systemd
-- Support: https://www.debian.org/support
-- 
-- An ExecStart= process belonging to unit stunnel4@webmin.service has exited.
-- 
-- The process' exit code is 'exited' and its exit status is 226.
Nov 10 22:11:10 ldap systemd[1]: stunnel4@webmin.service: Failed with result 'exit-code'.
-- Subject: Unit failed
-- Defined-By: systemd
-- Support: https://www.debian.org/support
-- 
-- The unit stunnel4@webmin.service has entered the 'failed' state with result 'exit-code'.
Nov 10 22:11:10 ldap systemd[1]: Failed to start Universal SSL tunnel for network daemons (webmin).
-- Subject: A start job for unit stunnel4@webmin.service has failed
-- Defined-By: systemd
-- Support: https://www.debian.org/support
-- 
-- A start job for unit stunnel4@webmin.service has finished with a failure.
-- 
-- The job identifier is 2473 and the job result is failed.
```

Hopefully this information is enough to get this solved.

**EDIT**: Just now saw [this](https://www.turnkeylinux.org/blog/webmin-1.960), and sadly the template fails to install completely in an unprivileged container:
```
Formatting '/storage/VMs/images/104/vm-104-disk-0.raw', fmt=raw size=8589934592
mke2fs 1.44.5 (15-Dec-2018)
Discarding device blocks:    4096/2097152               done                            
Creating filesystem with 2097152 4k blocks and 524288 inodes
Filesystem UUID: cd1633f1-298d-4bdd-a172-66a4f5ca2d83
Superblock backups stored on blocks: 
	32768, 98304, 163840, 229376, 294912, 819200, 884736, 1605632

Allocating group tables:  0/64     done                            
Writing inode tables:  0/64     done                            
Creating journal (16384 blocks): done
Multiple mount protection is enabled with update interval 5 seconds.
Writing superblocks and filesystem accounting information:  0/64     done

extracting archive '/storage/ISOs/template/cache/debian-10-turnkey-openldap_16.0-1_amd64.tar.gz'
tar: ./var/spool/postfix/dev/urandom: Cannot mknod: Operation not permitted
tar: ./var/spool/postfix/dev/random: Cannot mknod: Operation not permitted
Total bytes read: 556687360 (531MiB, 93MiB/s)
tar: Exiting with failure status due to previous errors
TASK ERROR: unable to create CT 104 - command 'lxc-usernsexec -m u:0:100000:65536 -m g:0:100000:65536 -- tar xpf - -z --totals --one-file-system -p --sparse --numeric-owner --acls --xattrs '--xattrs-include=user.*' '--xattrs-include=security.capability' '--warning=no-file-ignored' '--warning=no-xattr-write' -C /var/lib/lxc/104/rootfs --skip-old-files --anchored --exclude './dev/*'' failed: exit code 2
```

If there isn't another way I guess I'll have to allow nesting on my privileged container.

**Comment by JedMeister** [comment]
Oops, I just realised that I wrote this ~7 hours ago, but didn't click the button... Doh! Anyway, here it is...

@IzaacJ Thanks for your bug report.

As you've noticed, we have (hopefully!) resolved the issue in the latest Webmin package (which is still in our "testing" repo, but I hope to move to "main" repo really soon).

The other issue that you note (not being able to install our OpenLDAP appliance into an unprivileged container) is also a known issue. All appliances suffered from that in our v15.x release.

You can read all about that issue here -> #855 <- if you wish. But the TL;DR of it is that by default, Postfix runs within a chroot (`/var/spool/postfix/`), but to integrate with OpenLDAP, it requires those 2 files (`/var/spool/postfix/dev/urandom` & `/var/spool/postfix/dev/random`).

We resolved it in v16.0 appliances by simply removing those files when we build the containers as ee figured that most users won't miss the OpenLDAP integration. However, we also assumed that OpenLDAP integration for Postfix was likely a desirable thing, so we left it there for the OpenLDAP appliance.

So if you don't want/need OpenLDAP integration with Postfix, then you can follow the notes on how to workaround those 2 files in this comment -> https://github.com/turnkeylinux/tracker/issues/855#issuecomment-576103271 <-

Alternatively, if you do want to have Postfix integrated with OpenLDAP, then you don't have any option but to run in a privileged container. With our v16.x appliances, running as privileged, you'll hit another bug (new for Debian 10/Buster and inherited by TurnKey v16.x) #1452. The root of the issue seems to boil down to systemd service security hardening that has occurred in Debian which seems to be problematic within a privileged Buster LXC container (TurnKey) running on a Buster LXC (Proxmox) host. As noted in the OP of that bug, I note that the quickest, easiest workaround is to enabled "nested" mode on the Proxmox host. Although be aware that that has security implications, especially for a privileged LXC container.

**Comment by JedMeister** [comment]
I just read through your post again and actually, I'm pretty that the initial issue that you hit was was being caused by #1452.

But on the OpenLDAP appliance, we've left the Postfix OpenLDAP integration required files in place.

So in it's current state, it doesn't work either way... TBH, I'm not really sure on the best way forward with this one. I'd be really interested in your thoughts.

TBH, I'm leaning towards removing the `/var/spool/postfix/dev/urandom` & `/var/spool/postfix/dev/random` files...

What are your thoughts?

**Comment by IzaacJ** [comment]
Personally, I think leaving them out is the best approach. Requiring something to run in a priv'd container with nesting enabled feels like a bad idea.
Is it possible to remove postfix all together (in lxc builds) and let users chose to install it or not during the first configuration, along with a little notice?

**Comment by JedMeister** [comment]
> Personally, I think leaving them out is the best approach. Requiring something to run in a priv'd container with nesting enabled feels like a bad idea.

I'm mostly inclined to agree. Although at least in the current state, if the user has the inclination to report the issue (and/or do some research) then the limitation and trade-offs are known and the user can make an educated decision on which way they wish to go (or whether they'd rather just run in a "proper" VM - which "just works" and doesn't have any of these limitations or potential security trade-offs).

Also, to clarify; the requirement of running in a privileged container with nesting enabled only applies if you want to have OpenLDAP authentication for Postfix - without making config changes from the host. If you don't want/need the Postfix integration, then you can use the aforementioned workaround to remove the 2 special files and start the container as unprivileged and all should be well. As noted on the issue, there are alternate ways to work around it too (but require configuration on the host, that we can't apply from within the container), or you can just install from ISO to a KVM VM...

> Is it possible to remove postfix all together (in lxc builds) and let users chose to install it or not during the first configuration, along with a little notice? That might be a better solution.

It is, but that doesn't really resolve the issue. The issue isn't whether Postfix is installed or not, it's whether those 2 special files exist or not. If you install Postfix on an unprivileged container, then those 2 files are simply not created (the fact that they weren't created should appear in the dpkg logs, but the user will likely not realise unless they go hunting).

And to ensure we're on the same page, all other Postfix functionality works fine whether those 2 special files exist or not. It's only the LDAP integration that breaks without them (and it doesn't break in a way that would lead the user straight to the issue...).

So a user who wants LDAP integration with Postfix, who is unaware of this scenario and who starts an unprivileged container will need to change to a privileged container (or make the required host-side config changes) to make it work (plus replace those files).

So with your feedback and some more thoughts, here's what I think we should do:

- Remove the 2 special files from the Postfix chroot for the OpenLDAP appliance (as we do with all other v16.x appliances).
- Open a separate issue noting the limitations in different scenarios and the associated workarounds and trade-offs.
- Make a firstboot script to create the 2 special Postfix chroot files files if it has permission (so the Postfix part will "just work" when running privileged - the general issues of running Buster LXC on Buster will still exist).
- Also include a note on firstboot of our LXC builds that could detect whether it's running privileged or unprivileged and provide an appropriate message (pointing to the issue noted above).

Unfortunately, I'm not sure when we'll be able to get around to all that, but it'd be nice...

**Comment by timdonovanuk** [comment]
I found this via the issue I submitted here https://github.com/turnkeylinux/tracker/issues/1639. This is all way over my head but just wanted to note that using a privileged container with nesting enabled is a requirement of some FOSS projects, as they were built to support docker but also need direct access to hardware (for machine learning acceleration). 

Running these projects in docker on VM introduces a lot of headache passing through hardware.
